src/plugins/output_format/jsonfmt.h

```diff
@@ -0,0 +1,315 @@
+        if (print_data(os, fmt::qstr(data.first), 0))
+        {
+            os << ':';
+            if (print_data(os, data.second, ','))
+            {
+                return true;
+            }
+        }
+        os.seekp(pos);
+        return false;
+    }
+};
+
+template <class T>
+struct data_printer<T, std::enable_if_t<is_iterable<T>::value, void>>
+{
+    static bool print(fmt::ostream& os, const T& data, char sep)
+    {
+        uint32_t i = 0;
+        bool printed = false;
+        for (const auto& v : data)
+        {
+            if (printed)
+                os << sep;
+            printed = print_data(os, v, sep);
+            ++i;
+        }
+        if (i > 0 && !printed)
+            os.unputc(sep);
+        return true;
+    }
+};
+
+template <class T>
+constexpr bool print_data(fmt::ostream& os, const T& data, char sep)
+{
+    return data_printer<T>::print(os, data, sep);
+}
+
+/**/
+
+template <class T, class... Ts>
+constexpr bool print_data(fmt::ostream& os, const T& data, const Ts& ... rest)
+{
+    constexpr char sep = ',';
+    bool printed = print_data(os, data, sep);
+    bool printed_rest = false;
+
+    if constexpr (sizeof...(rest) > 0)
+    {
+        if (printed)
+            os << sep;
+        printed_rest = print_data(os, rest...);
+        if (!printed_rest && printed)
+            os.unputc(sep);
+    }
+    return printed || printed_rest;
+}
+
+/**/
+
+inline void print_common_data(fmt::ostream& os, const char* plugin_name, drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    if (info)
+    {
+        std::optional<decltype(fmt::qstr(info->trap->name))> method;
+        if (info->trap->name)
+            method = fmt::qstr(info->trap->name);
+
+        print_data(os,
+                   keyval("Plugin", fmt::qstr(plugin_name)),
+                   keyval("Time", TimeVal{UNPACK_TIMEVAL(info->timestamp)}),
+                   keyval("PID", fmt::nval(info->attached_proc_data.pid)),
+                   keyval("PPID", fmt::nval(info->attached_proc_data.ppid)),
+                   keyval("TID", fmt::nval(info->attached_proc_data.tid)),
+                   keyval("UserName", fmt::qstr(USERIDSTR(drakvuf))),
+                   keyval("UserId", fmt::nval(info->proc_data.userid)),
+                   keyval("ProcessName", fmt::qstr(info->attached_proc_data.name)),
+                   keyval("Method", method)
+                  );
+    }
+}
+
+template<class... Args>
+void print(const char* plugin_name, drakvuf_t drakvuf, drakvuf_trap_info_t* info, const Args& ... args)
+{
+    fmt::out << '{';
+
+    bool printed = false;
+    if (info)
+    {
+        print_common_data(fmt::out, plugin_name, drakvuf, info);
+        printed = true;
+    }
+
+    if constexpr (sizeof...(args) > 0)
+    {
+        constexpr char sep = ',';
+        if (printed)
+            fmt::out << sep;
+        if (!print_data(fmt::out, args...))
+            fmt::out.unputc(sep);
+    }
+
+    fmt::out << "}\n";
+    fmt::out.flush();
+}
```

Comment:
Here you are basically re-implementing JSON grammar, duplicating the logic of `json-c` library which is already included in DRAKVUF. I like the idea of unified interface for printouts, but could we just use `json-c` in `jsonfmt.h` in order to construct JSON objects? This would ensure that what's constructed is a valid JSON.

Reply:
Yes, performance will be lower, but it can be done with json-c.

Reply:
It can be done in next iteration refactoring.